utildb/gen_deleted_accounts.go

```diff
@@ -59,100 +67,284 @@ func readAccounts(ch chan proxy.ContractLiveliness, deleteHistory *map[common.Ad
 
 // genDeletedAccountsTask process a transaction substate then records self-destructed accounts
 // and resurrected accounts to a database.
-func genDeletedAccountsTask(
-	tx *substate.Transaction,
-	processor *executor.TxProcessor,
-	ddb *substate.DestroyedAccountDB,
-	deleteHistory *map[common.Address]bool,
-	cfg *utils.Config,
-) error {
+func genDeletedAccountsTask(tx *substate.Transaction, processor executor.TxProcessor, cfg *utils.Config, log logger.Logger) ([]proxy.ContractLiveliness, error) {
 	ch := make(chan proxy.ContractLiveliness, channelSize)
 	var statedb state.StateDB
 	var err error
 	ss := substatecontext.NewTxContext(tx.Substate)
 
 	statedb, err = state.MakeOffTheChainStateDB(ss.GetInputState(), tx.Block, state.NewChainConduit(cfg.ChainID == utils.EthereumChainID, utils.GetChainConfig(cfg.ChainID)))
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	//wrapper
-	statedb = proxy.NewDeletionProxy(statedb, ch, cfg.LogLevel)
+	statedb = proxy.NewDeletionProxy(statedb, ch, log)
 
 	_, err = processor.ProcessTransaction(statedb, int(tx.Block), tx.Transaction, ss)
 	if err != nil {
-		return nil
+		return nil, nil
 	}
 
 	close(ch)
-	des, res := readAccounts(ch, deleteHistory)
-	if len(des)+len(res) > 0 {
-		// if transaction completed successfully, put destroyed accounts
-		// and resurrected accounts to a database
-		if tx.Substate.Result.Status == types.ReceiptStatusSuccessful {
-			err = ddb.SetDestroyedAccounts(tx.Block, tx.Transaction, des, res)
-			if err != nil {
-				return err
-			}
-		}
+
+	livelinessArr := make([]proxy.ContractLiveliness, 0)
+	for liveliness := range ch {
+		livelinessArr = append(livelinessArr, liveliness)
 	}
 
-	return nil
+	return livelinessArr, nil
 }
 
 // GenDeletedAccountsAction replays transactions and record self-destructed accounts and resurrected accounts.
+// Uses round-robin task assignment system to workers to keep order while utilizing parallelism.
 func GenDeletedAccountsAction(cfg *utils.Config, ddb *substate.DestroyedAccountDB, firstBlock uint64, lastBlock uint64) error {
-	var err error
-
-	err = utils.StartCPUProfile(cfg)
+	err := utils.StartCPUProfile(cfg)
 	if err != nil {
 		return err
 	}
 
 	log := logger.NewLogger(cfg.LogLevel, "Generate Deleted Accounts")
-
 	log.Noticef("Generate deleted accounts from block %v to block %v", firstBlock, lastBlock)
 
-	start := time.Now()
-	sec := time.Since(start).Seconds()
-	lastSec := time.Since(start).Seconds()
-	txCount := uint64(0)
-	lastTxCount := uint64(0)
-	var deleteHistory = make(map[common.Address]bool)
+	processor := executor.MakeTxProcessor(cfg)
+
+	wg := sync.WaitGroup{}
+	stopChan := make(chan struct{})
```

Comment:
Have a look at [utils/event.go](https://github.com/Fantom-foundation/Aida/blob/main/utils/event.go) please. It's an alternative to using `stopChanels` implemented by Herbert. iirc he said its a functionality from `cpp`.

Reply:
Thanks, that's great alternative.